arch/risc-v/src/rv64gc/riscv_mmu.c

```diff
@@ -0,0 +1,107 @@
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+#include <arch/csr.h>
+
+#include "riscv_internal.h"
+#include "riscv_mmu.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void mmu_ln_setentry(uint32_t ptlevel, uint64_t lnvaddr, uint64_t paddr,
+                     uint64_t vaddr, uint32_t mmuflags)
+{
+  uint64_t *lntable = (uint64_t *)lnvaddr;
+  uint32_t  index;
+
+  DEBUGASSERT(ptlevel <= RV_MMU_PT_LEVELS);
+
+  /* Test if this is a leaf PTE, if it is, set A+D even if they are not used
+   * by the implementation.
+   *
+   * If not, clear A+D+U because the spec. says:
+   * For non-leaf PTEs, the D, A, and U bits are reserved for future use and
+   * must be cleared by software for forward compatibility.
+   */
+
+  if (mmuflags & PTE_LEAF_MASK)
+    {
+      mmuflags |= (PTE_A | PTE_D);
+    }
+  else
+    {
+      mmuflags &= ~(PTE_A | PTE_D | PTE_U);
+    }
+
+  /* Make sure the entry is valid */
+
+  mmuflags |= PTE_VALID;
+
+  /* Calculate index for lntable */
+
+  index = (vaddr >> RV_MMU_VADDR_SHIFT(ptlevel)) & RV_MMU_VPN_MASK;
+
+  /* Move PPN to correct position */
+
+  paddr >>= RV_MMU_PTE_PPN_SHIFT;
+
+  /* Save it */
+
+  lntable[index] = (paddr | mmuflags);
+
+  /* Update with memory by flushing the cache(s) */
+
+  mmu_invalidate_tlb_by_vaddr(vaddr);
+}
+
+void mmu_ln_map_region(uint32_t ptlevel, uint64_t lnvaddr, uint64_t paddr,
+                       uint64_t vaddr, size_t size, uint32_t mmuflags)
+{
+  uint64_t end_paddr = paddr + size;
```

Comment:
Do you need to somehow check or (debug)assert address alignment to page boundary?

Reply:
LGTM

Reply:
You could assert / check the validity of any of the parameters, but if they contain garbage it means there is a bug in the kernel which must be fixed anyway. Would not be harmful *if* it can be disabled (two separate build types, debug and release).